Kubernetes discovery

jimmidyson opened the pull request:
As discussed in #904. This initial implementation adds in node discovery which scrapes pod/container metrics from all nodes in the cluster.

Would love feedback please. Missing tests, although I guess only place that I can add them easily is to the config parsing? Oh & documentation.

Sorry about the amount of godeps required - not sure of how to avoid that if that's a problem? Certainly makes the PR harder to review. If you have a better way to add these deps in please let me know.

jimmidyson commented:
Added config test.

jimmidyson commented:
I've found that I've had to explicitly disable cgo with `CGO_ENABLED=0` in `GOENV` to be able to build including the kubernetes client without requiring `gcc`. I don't think this is a problem as I think everything else in Prometheus doesn't require gcc anyway?

jimmidyson commented:
Where would the appropriate place be for documentation? I'll wait for feedback on this PR & if it looks like getting accepted I'll work on docs in the prometheus.github.io repo?

beorn7 commented:
Oh wow... 616 files due to dependecies... CGO trouble, too...
I'll definitely have to discuss this with @fabxc and/or @juliusv . Both are on vacation right now. Stay tuned...

jimmidyson commented:
Ha yeah slightly bloated. I'm working on swapping to straight http.Client. Should do the job if I can figure out how to handle streamed chunked responses. Any tips?

jimmidyson commented:
@beorn7 I've switched to using http.Client & handling JSON responses via a `json.Decoder`. This works, but doesn't reduce the dependencies very much (~500 still). This is because the Kubernetes API packages require so many other packages... My only option if these dependencies are a problem is to duplicate the required types from the Kubernetes API packages into Prometheus - I'm a little bit reluctant to do that as it makes it harder to maintain & more fragile if the Kubernetes API changes. What do you think?

beorn7 commented:
Yeah, unfortunately a known and repeating problem... This requires a bit of thinking.

grobie commented:
Maybe worth to raise the issue of many dependencies to use types in
kubernetes itself?
On Jul 20, 2015 8:30 AM, "Björn Rabenstein" <notifications@github.com>
wrote:

> Yeah, unfortunately a known and repeating problem... This requires a bit
> of thinking.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/prometheus/prometheus/pull/905#issuecomment-122869867>
> .
>

jimmidyson commented:
@grobie Good idea. Raised an issue with kubernetes (GoogleCloudPlatform/kubernetes#11606), although not sure what they will be able to do about it, at least in the short term.

jimmidyson commented:
There's WIP to make the Kubernetes v1 API safer to import - see GoogleCloudPlatform/kubernetes#11280. I'll keep an eye on that & hope to see something to use in the next week or so. I'll switch to straight `http.Client` from the Kubernetes client to reduce package dependencies. Had this working earlier so simple enough to switch back.

jimmidyson commented:
@grobie Swapped to native `http.Client` now & it's reduced package dependencies significantly, but still a lot. The only package import I have to kubernetes is `github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1` & this generates this mass of godeps... Once they've fixed these tangled mess this should all still work (hopefully!) without any code changes, just dropping godeps.

juliusv commented:
@jimmidyson Great. Still can't review through GitHub since it limits the number of files it can display, but I'd prefer waiting for more of the Kubernetes dependency reductions anyways if that's ok for you. The internal SD interface in Prometheus should be pretty stable, so this PR shouldn't bit-rot too much.

jimmidyson commented:
@juliusv Totally agree about waiting for kubernetes to sort out their
package structure. I'll keep it up to date with any Prometheus changes as
required.

On 12:23, Wed, 22 Jul 2015 Julius Volz <notifications@github.com> wrote:

> @jimmidyson <https://github.com/jimmidyson> Great. Still can't review
> through GitHub since it limits the number of files it can display, but I'd
> prefer waiting for more of the Kubernetes dependency reductions anyways if
> that's ok for you. The internal SD interface in Prometheus should be pretty
> stable, so this PR shouldn't bit-rot too much.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/prometheus/prometheus/pull/905#issuecomment-123675110>
> .
>

jimmidyson commented:
Woohoo! No new dependencies.

Thanks for the suggestion on IRC @fabxc!

I think this is ready for review if anyone would be able to take a look.

jimmidyson commented:
FYI `retrieval/discovery/kubernetes/types.go` is basically a copy of https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1/types.go with all currently irrelevant types/fields stripped out. This removes the need for any kubernetes dependencies, with the drawback of having to keep this file up to date. This shouldn't be too bad though as the Kubernetes v1/types are now strictly versioned & stable.

fabxc commented on config/config.go:
```diff
@@ -508,6 +522,49 @@ func (c *ServersetSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) err
+	Server           string   `yaml:"server"`
+	NodeMetricsPort  int      `yaml:"node_metrics_port,omitempty"`
+	InCluster        bool     `yaml:"in_cluster,omitempty"`
+	BearerTokenFile  string   `yaml:"bearer_token_file,omitempty"`
```
All those authentication options are very similar to your recent PR for scrape authentication. I assume this is not specific to Kubernetes but because your organisation has the API secured in the same manner as your `/metrics` endpoints?

This would certainly be interesting for other SD mechanisms too, e.g. Consul or serversets.
How about moving this into its own config block and allow plugging it in into other mechanisms more easily?

If so, it would be easiest to remove it from this PR for now and add this it in a later one.

jimmidyson replied:
In a Kubernetes cluster, the masters are secured in many possible ways. I've implemented the 3 most commonly used: bearer tokens (easiest to use in cluster, automatically mounted bearer token files in pods via service accounts); basic auth; & mutual TLS. I could move that to common, but I'd rather do that in a separate PR tbh as without this it's impossible for metrics to be gathered from a Kubernetes cluster - authentication is required.

fabxc replied:
However you think is best.

Eventually having a `auth` configuration block that is common across several SD mechanisms would be a good thing to have.

jimmidyson replied:
Yeah. How about I refactor that after this PR is merged? Raise an issue & assign to me if you want ;)

fabxc replied:
Sounds good.

fabxc commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,582 @@
+	if err != nil {
+		// If we can't list nodes then we can't watch them. Assume this is a misconfiguration
+		// & panic rather than continue in a bad state
+		panic(fmt.Errorf("discovery.kubernetes.Sources: %s", err))
```
Panicking in malformed outside input is very risky. People can crash the server with a single tweak of the JSON input.
Let's log an error here and return the empty list. This is will only cause targets to disappear for a brief time on reloading - nothing inhibiting operation too much.

fabxc replied:
Have to invalidate the statement as `Sources()` succeeding is crucial in this implementation - still, logging an error is better than panicking.

fabxc commented on retrieval/target.go:
```diff
@@ -292,6 +289,10 @@ func newHTTPClient(cfg *config.ScrapeConfig) (*http.Client, error) {
 		rt = httputil.NewBearerAuthRoundTripper(bearerToken, rt)
 	}
 
+	if cfg.BasicAuth != nil {
```
Reading my mind much? 

:+1:

fabxc commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,582 @@
+	kubernetesServiceAccountToken  = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+	kubernetesServiceAccountCACert = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+
+	kubernetesApiVersion          = "v1"
```
Acronyms in uppercase. 

HTTP, API, URL ...

fabxc commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,582 @@
+	log.Debugf("Stopping Kubernetes discovery for %s", kd.conf.Server)
+
+	// The lock prevents Run from terminating while the watchers attempt
+	// to send on their channels.
```
This can hopefully be removed after the pipelining improvements in #977.
Don't worry about the changes over there for now - I'll adjust as necessary later.

jimmidyson replied:
OK cool - so no changes for this for now?

fabxc replied:
Yeah

fabxc commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,582 @@
+)
+
+const (
+	kubernetesSourcePrefix        = "kubernetes"
```
I mentioned in another PR that in the long run we probably want to move discovery mechanisms into their own packages. Maybe this is a good opportunity to start doing this so we can drop the prefixing altogether.

What do you think?

fabxc replied:
As you have to rebase you can drop the `kubernetesSourcePrefix` everywhere. You only have to ensure uniqueness within the scope a single instance of Kubernetes SD since the changes #970.

jimmidyson replied:
I'm happy to move this into it's own package if you'd prefer. Personally not too bothered by it.

fabxc replied:
The prefixing is usually fine for variables and constant, as long as it's done. But sooner or later internal helper functions can become messy and we probably don't want to prefix those.

I think having sub-packages in `retrieval/discovery/<sd_mechanism>` with the implementations and the constructors in `retrieval/discovery` would be nice.

fabxc commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,582 @@
+	kubernetesServiceNamespaceLabel = clientmodel.MetaLabelPrefix + "kubernetes_service_namespace"
+	// KubernetesServiceNameLabel is the name for the label containing a target's service name.
+	kubernetesServiceNameLabel = clientmodel.MetaLabelPrefix + "kubernetes_service_name"
+	// KubernetesNodesTargetGroupName is the name given to the target group for nodes
```
All comments `.`-terminated. Applies to several comments.

fabxc commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,582 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.package discovery
```
Remove package statement.

fabxc commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,582 @@
+		// Load CA cert
+		caCert, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			log.Fatal(err)
```
Regardless of whether or not this goes into this PR (see comment above) this constructor should return rather than log an error.

fabxc commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,582 @@
+
+	kd.updateNodesTargetGroup(ch)
+
+	for _, ns := range kd.services {
```
So `Run()` is implemented based on the assumption that `Sources()` was called before?
The other implementations do not assume this, IIRC. But it is in fact how it works. I'm wondering whether it's worth adding this as an invariant to the interface.

Thoughts?

brian-brazil replied:
The initial serverset SD depended on that assumption too.

Maybe an initial snapshot (not just sources but the targets too), and then updates from there might make more sense?

jimmidyson replied:
tbh I wasn't quite sure what the purpose of `Sources` was but implemented anyway & thought why not use it in `Run`... So what is the purpose of `Sources`?

fabxc replied:
`Sources()` was essentially the idea of a lightweight snapshot, i.e. not all SD mechanism (but many as I realize now) need to query the full data set to provide the sources.
I think there's no need to drop the idea of a lightweight snapshot but adding the mentioned assumption seems to make sense.

fabxc replied:
If we reload the configuration we stop the whole service discovery (as diffing configs and modifying all running SDs is absolutely impractical). We then start service discovery as configured again.
The targets, which are associated with a _source_, remain though. After starting the service discovery mechanisms we scan through all the sources and drop the targets associated with the ones that disappeared (either because the entire SD config was dropped or modified so that certain groups of targets no longer exist).

Does that make sense? @jimmidyson

brian-brazil replied:
> I think there's no need to drop the idea of a lightweight snapshot but adding the mentioned assumption seems to make sense.

I think there is, as there could be races due to internal state changing between the `Sources` and the `Run`.

jimmidyson replied:
Due to the dynamic nature of kubernetes there's very likely to be changes between `Sources` & `Run` - in fact totally possible that anything discovered in`Sources` is no longer valid after config reloaded & `Run` started again.

fabxc replied:
Makes sense. But doing a full snapshot does not change that.
The basic idea is that you initially tell all your sources on a call to `Sources()` and from then on are responsible to inform about those sources through the channel from `Run()`. This includes a) one initial target group update per source b) any subsequent changes.

Possibly this can be stated more clearly in the interface documentation.

fabxc replied:
So in that regard the way you implemented it makes a lot of sense for this type of service discovery. We just have to state in the interface that `Sources()` is guaranteed to be called before each call to `Run()`.

fabxc commented on config/config.go:
```diff
@@ -508,6 +522,49 @@ func (c *ServersetSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) err
+	CertFile         string   `yaml:"cert_file,omitempty"`
+	KeyFile          string   `yaml:"key_file,omitempty"`
+	CAFile           string   `yaml:"ca_file,omitempty"`
+	RetryDelay       Duration `yaml:"retry_delay,omitempty"`
```
I prefer `RetryInterval` to get some consistency across configuration sections.

fabxc commented on config/config.go:
```diff
@@ -508,6 +522,49 @@ func (c *ServersetSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) err
+	CAFile           string   `yaml:"ca_file,omitempty"`
+	RetryDelay       Duration `yaml:"retry_delay,omitempty"`
+	RequestTimeout   Duration `yaml:"request_timeout,omitempty"`
+	DiscoverServices bool     `yaml:"discover_services,omitempty"`
```
We thought about allowing a way to grab all services from Consul too, just couldn't decide on how to configure it 100%. A boolean probably works fine (again, consistency is key - so this would also be a decision for Consul).
Consul however allows for a list of explicit services too.

Technically this is another case that can be solved with relabeling, but is common enough to allow direct configuration.
Do you think it would make sense in the context of Kubernetes too?

brian-brazil replied:
I think we should start out with everything, and if there's a need to filter further we can consider it then.

jimmidyson replied:
The problem is that not all services will be scrapeable. This discovery requires that services are annotated with configurable annotations to enable scraping. Enabling this for everything by default will only increase noise in a heterogeneous env IMO.

This flag is actually there to disable service scraping totally so you only scrape nodes & get container (crgoup) metrics, rather than app level metrics.

fabxc replied:
Yes, can be added. Let's do it like this.

Just that we decided to do the opposite for Consul. I imagine most people starting out with a minor fraction of application being instrumented with Prometheus. But Kubernetes users might indeed be more likely to make heavy use of labels here.

Good idea about the `ScrapeKey` by the way.

brian-brazil replied:
> This discovery requires that services are annotated with configurable annotations to enable scraping. 

The serverset SD will scrape everything unless you configure relabels to tell it otherwise.

> scrape nodes & get container (crgoup) metrics, rather than app level metrics.

It sounds to me like there's 3 different SD mechanisms here.

jimmidyson replied:
@brian-brazil We only update the target group with annotated services so they won't be scraped without annotations.

Why 3 different ones? It's very convenient to do this all in one, simpler config, reusing API server watches, etc.

brian-brazil replied:
One for nodes, one for cgroups and one for jobs. Each are very different things, it'd be like bundling the node_exporter, cadvisor and something instrumented client_java into one when in reality they're three very distinct layers of infrastructure.

jimmidyson replied:
Well let's say 2 - cadvisor is used for everything here no need for node_exporter.

The problem is that Kubernetes components mix cgroup metrics for pods/containers in with the kubernetes components' (master, kubelet) app metrics (e.g. API response times). Not much we can do about that without scraping twice & relabelling separately - but that would be a bit overkill of course.

Hence we bundle both discovery mechanisms together...

brian-brazil replied:
@jimmidyson Thanks for the explanation. Why do you need the boolean here then? Relabelling can handle that.

fabxc commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,593 @@
+	for {
+		select {
+		case <-kd.runDone:
+			close(update)
```
Sticking those into `defer`s earlier on wouldn't hurt. (Also my consistency-itch :)

jimmidyson commented:
@fabxc Addressed all comments, although seem to have lost one from @brian-brazil who was discussing the relative merits of this as one or two service discovery service mechanisms. Not sure where that discussion has gone?

jimmidyson commented on config/config.go:
```diff
@@ -584,18 +598,57 @@ type MarathonSDConfig struct {
+	CAFile           string   `yaml:"ca_file,omitempty"`
+	RetryInterval    Duration `yaml:"retry_interval,omitempty"`
+	RequestTimeout   Duration `yaml:"request_timeout,omitempty"`
+	DiscoverServices bool     `yaml:"discover_services,omitempty"`
```
@brian-brazil The reason for this is to enable/disable discovery of kubernetes services. With this enabled (default), we watch all services & check the scrape key to see whether it should be scraped or not. If users don't want to enable this & just collect cluster level metrics this can be disabled: this reduces the resources for 2 more watches, long polling connections to the kubernetes master.

brian-brazil replied:
I take it those are potentially expensive watches? Seems okay then.

jimmidyson replied:
They could cause a lot of change: as services are added/removed; as endpoints for those services are added/removed - this will cause target updates.

brian-brazil replied:
That's fine, they should be quite cheap to handle.

jimmidyson replied:
IMO this minor piece of functionality is worth the reduction in resource, both in Prometheus & Kubernetes, however cheap. Why do you think it's not worth adding?

brian-brazil replied:
We generally try to avoid adding functionality that's already possible with relabelling. This helps keep the number of configuration options and permutations sane.

jimmidyson replied:
@brian-brazil Fair enough - removing.

jimmidyson commented:
Just seen one thing to do with target group names I want to fix up after the rebase. Will fix up tomorrow.

brian-brazil commented on config/config.go:
```diff
@@ -584,18 +598,57 @@ type MarathonSDConfig struct {
+	RetryInterval    Duration `yaml:"retry_interval,omitempty"`
+	RequestTimeout   Duration `yaml:"request_timeout,omitempty"`
+	DiscoverServices bool     `yaml:"discover_services,omitempty"`
+	ScrapeKey        string   `yaml:"scrape_key,omitempty"`
```
These key options all seem to be doing a more primitive version of relabelling, and you're hardcoding that the user has a particular deployment where particular labels must be set. Anything we provide with Prometheus needs to be generic enough to work for all reasonable deployment models.

It's not clear to me that this supports a "scrape everything" model, and scheme/port/path should be done with relabelling.

brian-brazil commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,593 @@
+		},
+	}
+
+	for k, v := range service.ObjectMeta.Labels {
```
You're assuming here that the person running the prometheus is the same person setting the labels, if they're different then it's not safe to allow people override the labelling wishes of the person running the prometheus. These should be prefixed under __meta. The label names also need to be sanitized.

brian-brazil replied:
Thinking a bit, there's also a problem here as Kubernetes allows changing of labels after an object is created. Target labels should be constant over the life of the target to avoid all the timeseries changing names, as that leads to discontinuities in graphs and can cause other problems.

matthiasr replied:
Changing the labels on an object is meant to be a discontinuity, I think reflecting that is the right thing to do. If e.g. you want to measure request rates for a service, aggregate over all the labels that the service used to address it matches. The underlying set of pods changing, or a pod existing on but leaving the set, are expected and should be reflected.

Other problems are a bug.

brian-brazil replied:
> Changing the labels on an object is meant to be a discontinuity, I think reflecting that is the right thing to do

That depends on exactly what you're using each of your labels for. Target labels changing over the lifetime of the target is something that requires very careful consideration on a case-by-case basis.

brian-brazil replied:
> Reusing kubernetes labels as Prometheus labels is really the only sane way to do it - meta & relabelling would just be too annoying & arduous.

That's not safe or manageable for the user in the general case, particularly for Kubernetes which encourages such dynamic labels for any purpose. If you do this then there is no way for the user to undo it via relabelling (as they don't know a-priori what labels are present), so this needs to be opt-in on a per-label basis. Only the user knows which labels are safe to end up on the target in their particular deployment, and can ensure that they won't clash with the labels used by their applications.

> We've discussed this in the marathon PR IIRC

All I see on that PR are moving of such labels into meta labels, and removal of deployment-specific code. #979 filed for removal of remaining deployment-specific code.

> Changing kubernetes labels & creating discontinuities is fine: labels are the only thing we can use to tie resources together & reflect changes.

Even for a label called `brian_testing_please_ignore`? We need to design for the most general case, and there being a single label called `environment` is not the most general case.

> How do I sanitise labels?

Usually by replacing anything that doesn't match https://github.com/prometheus/client_golang/blob/master/model/labelname.go#L74 with an underscore. We should really add a utility function for that as that code has gotten to a lot of places by now.

jimmidyson replied:
> All I see on that PR are moving of such labels into meta labels, and removal of deployment-specific code. #979 filed for removal of remaining deployment-specific code.

The discussion was with @fabxc on an outdated diff - https://github.com/prometheus/prometheus/pull/902#discussion_r36508629. Happy to discuss this again though, but the same argument applies.

How about a label whitelist that could be a slice of regexes? All non-matching would go to meta labels, otherwise "proper" labels? That way users could use `.*` to pass through all labels. Default to making all meta labels? I would almost definitely be using this to whitelist everything of course :)

> We need to design for the most general case, and there being a single label called environment is not the most general case.

I'm not sure I follow. What do you mean by general case? IMO in this case users are not going to want to label everything separately, relabel extensively & correlate those labels to resources somehow. Reusing Kubernetes labels seems to be the most general case I can think of. This is a (potentially multi-tenant) environment where Prometheus can be provided as a drop-in service with very little configuration. To get this being used & gather further community feedback to improve I would suggest going with this 1-1 label mapping to begin with.

brian-brazil replied:
> How about a label whitelist that could be a slice of regexes? All non-matching would go to meta labels, otherwise "proper" labels?

That could be an option if it's completely opt-in, but we're lacking all the other support needed to make that manageable for the user on the other end as we presume you know the names of the labels you want to work with. There's a lot of scope here for shooting yourself in the foot, as non-constant target labels are a big anti-pattern. In the extreme you end up with every gauge as a label on the target - which sounds useful for grouping but doesn't work out in practice at all.

The usual way to solve this is to have a single timeseries with all these labels and the value of 1 that you can join to things. Then the rest of your timeseries don't break, and you can still make use of the labels. This'd be the first time that'd be coming from SD rather than the target itself so that'd need plumbing.

> I'm not sure I follow. What do you mean by general case? IMO in this case users are not going to want to label everything separately, relabel extensively & correlate those labels to resources somehow. 

Labels in Kubernetes aren't just for resource management, they're also for thing like selecting what goes into a load balancer. One potential use case is that every client of your service had it's own load balancer for each client, with it's own label for selecting pods. That'd completely break monitoring with the setup you propose here as the target labels could change every few minutes.

brian-brazil replied:
Thinking on it, there's other SD this will also come up and it is a plausible configuration that a user has reserved a prefix within their label namespace to automatically appear in Prometheus.

Given that, all SDs where this is possible (Kubernetes, Marathon, EC2 offhand) should put labels under `__meta` and we should add a new relabel action that'd take a regex, apply it to all labels names and all matching labels are copied to the replacement label name.

jimmidyson replied:
> There's a lot of scope here for shooting yourself in the foot, as non-constant target labels are a big anti-pattern.

Service labels (which this uses) tend to be fairly constant, it is labels on pods, RCs, etc that tend to change more often from what I've seen.

> In the extreme you end up with every gauge as a label on the target

The gauge isn't part of the Kubernetes label set, that comes from the scraped metric, right?

> One potential use case is that every client of your service had it's own load balancer for each client, with it's own label for selecting pods.

That sounds like a special case, not general case, to me. Also, they would necessarily be separate services with separate selectors & would want to get metrics on those separately, as well as aggregating over all related service types. Not sure why this wouldn't work correctly, other than the change every few minutes, which is a special case, not general case IMO.

If we did go with meta labels how would you propose that works for relabelling?

jimmidyson replied:
> Given that, all SDs where this is possible (Kubernetes, Marathon, EC2 offhand) should put labels under __meta and we should add a new relabel action that'd take a regex, apply it to all labels names and all matching labels are copied to the replacement label name.

I like that idea. So we would have to match label names with a regex, copy to replaced name. Could add that as separate PR if it's agreed to go with this approach & update this SD to only use meta labels.

brian-brazil replied:
> Service labels (which this uses) tend to be fairly constant, it is labels on pods, RCs, etc that tend to change more often from what I've seen.

I wasn't aware of the distinction. Will this give access to the pods labels too?

> The gauge isn't part of the Kubernetes label set, that comes from the scraped metric, right?

Yeah, people have done things like try and add label a for each service running on a machine to node exporter. I see this as the same, but at a different level.

> That sounds like a special case, not general case, to me. Also, they would necessarily be separate services with separate selectors & would want to get metrics on those separately, as well as aggregating over all related service types. Not sure why this wouldn't work correctly, other than the change every few minutes, which is a special case, not general case IMO.

The general case is being able to handle all of the special cases. We need to ensure that the default setup works (and labels changing every few minutes isn't working) for all reasonable use cases, while also allowing for specific use cases by additional configuration.

> Could add that as separate PR if it's agreed to go with this approach & update this SD to only use meta labels.

I can do that up as a separate PR for you.

jimmidyson replied:
> I wasn't aware of the distinction. Will this give access to the pods labels too?

Not for application level metrics, no. Only service labels due to their less volatile nature. Could add pod labels in a future iteration if they become necessary, but I see them as less useful from a service monitoring perspective. The pod address is available so bad performing pods can be identified & managed.

> The general case is being able to handle all of the special cases.

Of course you are right - what was I thinking?!

Agreed - I'll switch to meta labels & you (or I) can add in regex relabelling. For the meta labels, happy to use a prefix of `__meta_kubernetes__` or do you think should just go with `__meta__`? Want to be able to differentiate between "normal" meta labels & ones discovered from Kubernetes labels.

brian-brazil replied:
> Of course you are right - what was I thinking?!

General and typical are often confused.

> For the meta labels, happy to use a prefix of __meta_kubernetes__ or do you think should just go with __meta__?

All your meta labels should be `__meta_kubernetes_something`, so `__meta_kubernetes_labels_XXX` or similar. We could have `__meta_kubernetes_podlabels_XXX`in the future.

#980 has the relabel PR.

jimmidyson commented on retrieval/discovery/kubernetes.go:
```diff
@@ -0,0 +1,589 @@
+		},
+	}
+
+	for k, v := range service.ObjectMeta.Labels {
```
We need to be able to correlate kubernetes labels & Prometheus labels in order to build queries & display resources together. This is pretty fundamental. We've discussed this in the marathon PR IIRC. Reusing kubernetes labels as Prometheus labels is really the only sane way to do it - meta & relabelling would just be too annoying & arduous.

Changing kubernetes labels & creating discontinuities is fine: labels are the only thing we can use to tie resources together & reflect changes.

How do I sanitise labels?

jimmidyson commented:
@fabxc Moved to `discovery/kubernetes` package - much prefer the removal of prefix from var names.

@brian-brazil Switched to meta labels.

brian-brazil commented on config/config.go:
```diff
@@ -109,6 +109,17 @@ var (
+
+	// The default Kubernetes SD configuration
+	DefaultKubernetesSDConfig = KubernetesSDConfig{
+		NodeMetricsPort: 10255,
```
Where does this number come from?

jimmidyson replied:
Kubelet port currently listens on port 10250, but is secured by mutual TLS to the Kubernetes master. There is currently no way to auth to this without managing certificates to talk to kubelet. So in Kubernetes & OpenShift the current way to manage this is to open a read-only port on the kubelet - this is done by default installations on port 10255, hence this port number here. Once auth to the kubelet is sorted properly in kubernetes/openshift, this port will change to 10250 & use whatever mechanism they put in place for auth.

brian-brazil replied:
If that's a well-known port number then that's okay, the name could be clearer though as I'd presume by default that this applied to all discovered targets.

jimmidyson replied:
Don't know how to be clearer that `NodeMetricsPort` to say "this is the port that metrics from the node are collected from" :)

brian-brazil replied:
`KubeletPort` would be clearer. `NodeMetricsPort` could be read as the node exporter, something to do with node.js and as anything else that could be considered a node.

jimmidyson replied:
That is, indeed, much clearer - thanks.

jimmidyson replied:
I'm too blinkered on Kubernetes to think of the bigger picture - glad for your review.

brian-brazil commented on config/config.go:
```diff
@@ -109,6 +109,17 @@ var (
+	// The default Kubernetes SD configuration
+	DefaultKubernetesSDConfig = KubernetesSDConfig{
+		NodeMetricsPort: 10255,
+		ScrapeKey:       "prometheus.io/scrape",
```
These 4 options should be handled via relabelling.

jimmidyson replied:
What do you mean? These are annotations on the kubernetes service to be able to configure the targets via discovery. They do that by setting labels on the targets (relabelling) during discovery phase, but this needs to be configurable to allow certain kubernetes providers to change this key if they want to.

brian-brazil replied:
These are your annotations, not everyones. To be generic we should allow this to be configurable by relabelling.

jimmidyson replied:
What is the difference in doing this via relabelling or via SD config? Not sure. This isn't specific to my setup, but a pre-configured key to be used on annotations on service to enable scraping of service endpoints. I'm totally unsure how relabelling could be used in this case as this annotation actually prevents target groups from being created if the key isn't present & truthy.

brian-brazil replied:
> What is the difference in doing this via relabelling or via SD config?

We want to generally avoid every SD mechanism having to reimplement the same things again and again. We also want to avoid SD mechanisms that make assumptions about all deployments that aren't actually true.

Thus we have relabelling, as it turns out almost everything that SD configs want to do can be handled by it and that most of those things weren't globally applicable in the first place. The most a SD should be providing is a host, port and a pile of metadata.

> This isn't specific to my setup, but a pre-configured key to be used on annotations on service to enable scraping of service endpoints. 

Where is this pre-configured key coming from? Is it guaranteed to be used this way for all Kubernetes setups in existence? If it is, why isn't it following the existing naming scheme for these labels? Where was the discussion to create this convention?

> I'm totally unsure how relabelling could be used in this case as this annotation actually prevents target groups from being created if the key isn't present & truthy.

The `drop` and `keep` relabel actions allow for limiting what targets are created.

jimmidyson replied:
Ah OK so you're saying the better way to handle this is to pass through service annotations as well & use target relabelling to keep/drop based on that. Right I get it now, thanks! I agree this adds much more flexibility & also reduces the complexity of the Kubernetes discovery piece.

So no filtering in the discovery piece, pass it all through & let relabelling drop/keep targets.

Is my understanding correct now?

brian-brazil replied:
> Ah OK so you're saying the better way to handle this is to pass through service annotations as well & use target relabelling to keep/drop based on that. 

Yes.

> I agree this adds much more flexibility & also reduces the complexity of the Kubernetes discovery piece.

It does have the disadvantage that the prometheus configs are longer, but my expectation is that users were going to copy&paste from documentation anyway.

brian-brazil commented on documentation/examples/prometheus-kubernetes.yml:
```diff
@@ -0,0 +1,33 @@
+  - server: 'https://kubernetes.default.svc'
+    in_cluster: true
+
+  metric_relabel_configs:
```
I don't think you mean to use metric relabelling here

jimmidyson replied:
What did I mean to use? I'm still a Prometheus n00b (as you probably guessed ;)) so all advice welcome.

brian-brazil replied:
`relabel_configs`, as they affect target groups.

The metrics relabel configs affect the actual scrape (but not things like `up`) and are typically used as a tactical option to drop a big metric on the floor.

jimmidyson replied:
OK that makes sense. Thanks.

jimmidyson replied:
Now I revisit this & remember what I had done, `metric_relabel_configs` is correct. These metrics are scraped from the kubelet, a single target exposing all cadvisor metrics for all containers running on that node. So this has to be done in metric relabelling sadly.

brian-brazil replied:
This is going to affect all scraped kubernetes metrics, so you'll need to be much more specific on the source labels to only affect the right target. In fact given that this is a regex applied to every ingested sample, you may want a separate scrape config for it.

I don't suppose this is something the kublet could do itself?

jimmidyson replied:
I guess it is tbh. Let me see if I can get a PR into kubernetes for that.
Otherwise this is going to be pretty heavy on Prometheus.

On 13:34, Wed, 12 Aug 2015 Brian Brazil <notifications@github.com> wrote:

> In documentation/examples/prometheus-kubernetes.yml
> <https://github.com/prometheus/prometheus/pull/905#discussion_r36853872>:
>
> > +  scrape_timeout: 10s
> > +
> > +scrape_configs:
> > +# The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
> > +- job_name: 'prometheus'
> > +
> > +  target_groups:
> > +  - targets: ['localhost:9090']
> > +
> > +- job_name: 'kubernetes'
> > +
> > +  kubernetes_sd_configs:
> > +  - server: 'https://kubernetes.default.svc'
> > +    in_cluster: true
> > +
> > +  metric_relabel_configs:
>
> This is going to affect all scraped kubernetes metrics, so you'll need to
> be much more specific on the source labels to only affect the right target.
> In fact given that this is a regex applied to every ingested sample, you
> may want a separate scrape config for it.
>
> I don't suppose this is something the kublet could do itself?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/prometheus/prometheus/pull/905/files#r36853872>.
>

jimmidyson replied:
Dropped `metric_relabel_configs` & will get a PR into kubernetes to do it there.

brian-brazil commented on retrieval/discovery/kubernetes/discovery.go:
```diff
@@ -0,0 +1,587 @@
+			nodeLabel:                clientmodel.LabelValue(nodeName),
+		}
+		for k, v := range node.ObjectMeta.Labels {
+			t[clientmodel.LabelName(k)] = clientmodel.LabelValue(v)
```
This needs updating too

jimmidyson replied:
Good spot - thanks!

jimmidyson commented:
@brian-brazil Moved to a relabelling approach & it works great thanks! Still a few issues to sort out, but getting close now I think.

brian-brazil commented on documentation/examples/prometheus-kubernetes.yml:
```diff
@@ -0,0 +1,39 @@
+    in_cluster: true
+
+  relabel_configs:
+  - source_labels: [__meta_kubernetes_node,__meta_kubernetes_service_annotation_prometheus_io_scrape]
```
space after the , usually

brian-brazil commented on documentation/examples/prometheus-kubernetes.yml:
```diff
@@ -0,0 +1,39 @@
+
+scrape_configs:
+# The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
+- job_name: 'prometheus'
```
I'd remove this to avoid confusion

brian-brazil commented:
:+1: from me, @fabxc still needs to finish his review.

jimmidyson commented:
@brian-brazil Thanks for all your input - very much appreciated & I've learnt a lot too.

jimmidyson commented:
@brian-brazil We discussed a PR to kubernetes to tidy up the kubernetes metadata on metrics so we don't have to do metric relabelling. Just looked at source code - it embeds cadvisor & lets that register the container metrics with prometheus. Problem is that cadvisor doesn't know about any kubernetes metadata, just docker metadata, so the `name` label on the metric comes through with the name of the docker container. That is constructed from kubernetes metadata so contains everything relevant (namespace, pod name, container name), but in a hard to consume format after ingestion into prometheus.

Do you have any suggestions on how to handle that? I'd like to be able to rewrite the cadvisor metrics before they're exposed in the prometheus handler but can't see any way to do that.

brian-brazil commented:
@jimmidyson In that sort of case you have to do it via metric relabelling on the prometheus side. The key thing is that those rules only be applied to the cadvisor data, and not all the other targets too. So we'll be looking at having two scrape configs.

fabxc commented on retrieval/discovery/kubernetes/discovery.go:
```diff
@@ -0,0 +1,552 @@
+			return
+		default:
+		}
+		func() {
```
Just `f()` will do.

Also no reason to keep sleeping when we can receive from `stopCh`.
How about something like this?

```
f()
for {
    select {
    case <-stopCh:
        return
    case <-time.After(period):
        f()
    }
}
```

jimmidyson replied:
Doesn't really work for the behaviour I want: invoke `f` unless channel is already stopped (hence `select` first, checking `stopCh`) before invoking `f`.

jimmidyson replied:
Removing initial `f()` in your example would mean waiting for `period` for first invocation - again not desirable.

fabxc replied:
I see, then we have to do two `select` statements. More code, but still better then sleeping if we could already have shutdown.

fabxc commented on retrieval/discovery/kubernetes/discovery.go:
```diff
@@ -0,0 +1,552 @@
+	if err != nil {
+		// If we can't list nodes then we can't watch them. Assume this is a misconfiguration
+		// & log & return empty.
+		log.Errorf("discovery.Sources: %s", err)
```
The `<pkg>.<type>.<method>` style is what we are using for panics in many places. For actually logged errors let's have a better description.

Panics are for things than can be tracked down to implementation issues, thus the basic info relevant to developers. Those errors are all something that can actually happen during operation.

fabxc commented on retrieval/discovery/kubernetes/discovery.go:
```diff
@@ -0,0 +1,552 @@
+		sourceNames = append(sourceNames, sourceServicePrefix+":"+service.ObjectMeta.Namespace+"/"+service.ObjectMeta.Name)
+		namespace, ok := kd.services[service.ObjectMeta.Namespace]
+		if !ok {
+			namespace = make(map[string]*Service)
```
We only use `new` and `make` if absolutely necessary. So `map[string]*Service{}` is the agreed standard for Prometheus.

fabxc commented on retrieval/discovery/kubernetes/discovery.go:
```diff
@@ -0,0 +1,552 @@
+
+	kd.servicesResourceVersion = services.ResourceVersion
+	for idx, service := range services.Items {
+		sourceNames = append(sourceNames, sourceServicePrefix+":"+service.ObjectMeta.Namespace+"/"+service.ObjectMeta.Name)
```
This concatenation is happening in three places. A small helper like `serviceSource(service)` makes things simpler and easier to read.

fabxc commented on retrieval/discovery/kubernetes/discovery.go:
```diff
@@ -0,0 +1,552 @@
+	updatedNodeName := node.ObjectMeta.Name
+	switch eventType {
+	case "Deleted":
+		// Deleted - remove from nodes map
```
It's all about the dots ;)

fabxc commented:
LGTM :+1:

As discussed, this will be a beta-feature with the next release (along with Marathon SD). I understand you have this working in your environment and seem motivated to stick around for further improvements. So we don't have to review every single line to death.

Let's get some mileage on this.

jimmidyson commented:
Yeah I'm around to do my best to make this successful. I'll start on the docs soon.

pires commented:
Awesome @jimmidyson!!

jimmidyson commented:
Thanks @pires!

xiaoping378 commented:
Hi @jimmidyson,
So, kubernetes SD scrape mertics through the communication with kube-apiserver ? 
How it affects the k8s's performance?
have the plan of moving this SD to top of heapster ?

xiaoping378 commented:
yes, A single web request, but API server will response much data...
you think there is mirror effect ?

jimmidyson commented:
Quantify how much data? From the API server it isn't that much. Against my local API server: ~400kb, which I would consider very little. Note that Prometheus also scrapes nodes for container metrics which is likely to be larger but still well within safe limits.

You're more than welcome to profile what impact scraping has against API servers & nodes, but we've not seen any visible impact of this set up.

xiaoping378 commented:
Got it.
i just worry about it, no any quantify data ...
Thank you very much for your such a detailed reply.
continue to use prometheus, it give me a alert, other application's monitor.
